File: haskell/rover.hs
data Rover = Rover Int Int Direction deriving (Show)
data Direction = North|South|East|West deriving (Show)

left (Rover x y North) = Rover x y West
left (Rover x y West) = Rover  x y South
left (Rover x y South) = Rover x y East
left (Rover x y East) = Rover  x y North

right (Rover x y North) = Rover x y East  
right (Rover x y East) = Rover  x y South
right (Rover x y South) = Rover x y West 
right (Rover x y West) = Rover  x y North

move (Rover x y North) = Rover (x+1) y East 
move (Rover x y East) = Rover  x (y+1) South 
move (Rover x y South) = Rover x (y-1) West 
move (Rover x y West) = Rover  (x+1) y North 


r = right (move (left (move (left (Rover 1 2 North)))))

File: Java/fp/src/main/java/rover/v1/Rover.java
package rover.v1;

enum Direction{North,South,East,West}

public class Rover {
    int x;
    int y;
    Direction direction;

    public void left(){
        switch (direction){
            case North:
                direction=Direction.West;
                break;
            case West:
                direction=Direction.South;
                break;
            case South:
                direction=Direction.East;
                break;
            case East:
                direction=Direction.North;
        }

    }

    public void right(){
        switch (direction){
            case North:
                direction=Direction.East;
                break;
            case West:
                direction=Direction.North;
                break;
            case South:
                direction=Direction.West;
                break;
            case East:
                direction=Direction.South;
        }

    }

    public void move(){
        switch (direction){
            case North:
                y++;
                break;
            case West:
                x--;
                break;
            case South:
                y--;
                break;
            case East:
                x++;
        }

    }


}
